cmd/snap/cmd_run.go

```diff
@@ -91,6 +92,42 @@ and environment.
 		}, nil)
 }
 
+func maybeWaitForSecurityProfileRegeneration() error {
+	// check if the security profiles key has changed, if so, we need
+	// to wait for snapd to re-generate all profiles
+	mismatch, err := interfaces.SystemKeyMismatch()
+	if err != nil {
+		// something went wrong with the system-key compare, there
+		// is nothing we can do but continue and hope for the best
+		logger.Debugf("SystemKeyMismatch returned an error: %v", err)
+		return nil
+	}
+	if !mismatch {
+		return nil
+	}
+
+	// We have a mismatch, try to connect to snapd, once we can
+	// connect we just continue because that means snapd has
+	// generated new profiles.
+	//
+	// connect timeout is 5s
+	timeout := 12
+	if timeoutEnv := os.Getenv("SNAPD_DEBUG_SYSTEM_KEY_RETRY"); timeoutEnv != "" {
+		if i, err := strconv.Atoi(timeoutEnv); err == nil {
+			timeout = i
+		}
+	}
+
+	cli := Client()
+	for i := 0; i < timeout; i++ {
+		if _, err := cli.SysInfo(); err == nil {
+			return nil
+		}
```

Comment:
shouldn't we sleep a bit here?

Reply:
Yes, I think this makes sense, cli.SysInfo() should sleep when it cannot connect but we should also do it for good measure.